libs/ml/examples/bert_save_load_pretrained_model_example/main.cpp

```diff
@@ -0,0 +1,527 @@
+// bert model creation functions
+std::pair<std::vector<std::string>, std::vector<std::string>> LoadPretrainedBertModel(
+    std::string const &file_path, BERTConfig const &config, GraphType &g);
+TensorType RunPseudoForwardPass(std::vector<std::string> input_nodes, std::string output_node,
+                                BERTConfig const &config, GraphType g, SizeType batch_size,
+                                bool verbose = false);
+
+void      SaveGraphToFile(GraphType &g, std::string const file_name);
+GraphType ReadFileToGraph(std::string const file_name);
+
+int main(int ac, char **av)
+{
+  // the example takes in two arguments: first for pretrained bert txt folder, second for bin file
+  // to save model
+  // for the pretrained bert txt folder, you can generate this txt weight files with the relevant
+  // scripts in https://github.com/uvue-git/fetch-ledger-test-scripts
+  if (ac != 3)
+  {
+    std::cout << "Wrong number of / No available argument" << std::endl;
+    return 1;
+  }
+
+  // from and to dir
+  std::string pretrained_model_dir = av[1];
+  std::string saved_model_path     = av[2];
+
+  // load pretrained bert model and print its output of a toy input
+  BERTConfig    config;
+  BERTInterface interface(config);
+  GraphType *   g = new GraphType();
+
+  std::cout << "load pretrained pytorch bert model from folder: \n"
+            << pretrained_model_dir << std::endl;
+  LoadPretrainedBertModel(pretrained_model_dir, config, *g);
+
+  std::cout << "get an output for the bert loaded from txt files" << std::endl;
+  TensorType first_output =
+      RunPseudoForwardPass(interface.inputs, interface.outputs[interface.outputs.size() - 1],
+                           config, *g, static_cast<SizeType>(1), false);
+
+  std::cout << "save the pretrained bert model to file: \n" << saved_model_path << std::endl;
+  SaveGraphToFile(*g, saved_model_path);
+
+  // delete the model for memory efficiency
+  delete g;
+
+  std::cout << "load saved model for testing" << std::endl;
+  GraphType g2 = ReadFileToGraph(saved_model_path);
+
+  std::cout << "get another output for the bert loaded from bin file" << std::endl;
+  TensorType second_output =
+      RunPseudoForwardPass(interface.inputs, interface.outputs[interface.outputs.size() - 1],
+                           config, g2, static_cast<SizeType>(1), false);
+
+  if (first_output == second_output)
+  {
+    std::cout << "The saved model matched the origin model, congrats!!!" << std::endl;
+  }
+  else
+  {
+    std::runtime_error("The serialization is not working properly");
+  }
+
+  return 0;
+}
+
+TensorType LoadTensorFromFile(std::string file_name)
+{
+  std::ifstream weight_file(file_name);
+  assert(weight_file.is_open());
+
+  std::string weight_str;
+  getline(weight_file, weight_str);
+  weight_file.close();
+
+  return TensorType::FromString(weight_str);
+}
+
+void PutWeightInLayerNorm(StateDictType &state_dict, SizeType model_dims,
```

Comment:
can you split this into two or three files please all under examples/bert_save_load_pretrained_model_example/, for example:
examples/bert_save_load_pretrained_model_example/load_save_utilities.hpp
examples/bert_save_load_pretrained_model_example/bert_weight_setting_utilities.hpp
etc.

It would be nice if main.cpp was just the main function, the bert config and maybe one or two other core components, these utility functions are large but auxiliary and make it harder to read what the actual logic flow is

Reply:
sure I will move them to utilities